### PR TITLE
TypeScript support for new Async API

### PR DESCRIPTION
**Pull request by christian-bromann**
With getting rid of fibers we want to support a better async API, e.g.:
```js
await browser.url('https://webdriver.io')
const tagName = await $('foo').$$('bar').get(2).getTagName()
const elem = await $('body').$('header').$$('div').find(
    async (elem) => {
        const loc = await elem.getLocation()
        return loc.x === 30
    }
)
console.log((await elem.getLocation()).x) // 30
console.log(await $('foo').$$('bar').length) // 5
```

This API needs to have proper type support.

### CODE REVIEW

File: packages/webdriverio/src/types.ts [hunk 1]
```diff
@@ -11,9 +11,78 @@ import type DevtoolsInterception from './utils/interception/devtools'
...
+interface ChainablePromiseArray<T> extends Promise<T> {
+    /**
+     * Amount of element fetched.
+     */
+    length: Promise<number>
+
+    /**
+     * Unwrap the nth element of the element list.
+     */
+    get: (i: number) => T
```
Comment (mgrybyk) [inline packages/webdriverio/src/types.ts:63]: I'd replace `get` method with accessing element by index in the elements array

File: packages/webdriverio/src/types.ts [hunk 2]
```diff
@@ -11,9 +11,78 @@ import type DevtoolsInterception from './utils/interception/devtools'
...
+    forEach: (iterator: (elem: T) => void) => ChainablePromiseArray<T>
+    forEachSeries: (iterator: (elem: T) => void) => ChainablePromiseArray<T>
```
Comment (mgrybyk) [inline packages/webdriverio/src/types.ts:65]: What do *Series methods stand for?

File: packages/webdriverio/src/types.ts [hunk 3]
```diff
@@ -11,9 +11,78 @@ import type DevtoolsInterception from './utils/interception/devtools'
...
+    findSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    findIndex: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    findIndexSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    some: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    someSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    every: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    everySeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    filter: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    filterSeries: (iterator: (elem: T) => boolean) => ChainablePromiseArray<T>
+    reduce: (iterator: (initialValue: any, elem: T) => any, initialValue: any) => any
```
Comment (mgrybyk) [inline packages/webdriverio/src/types.ts:78]: Origina Array's `reduce` method definition looks like this 
```
reduce<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number, array: T[]) => U, initialValue: U): U;
```
Is it possible to adjust our wrapper accordingly?

File: packages/webdriverio/src/types.ts (hunk 2 above)
Comment (christian-bromann) [reply packages/webdriverio/src/types.ts:65]: same as the companion command but runs only one callback at a time.

File: packages/webdriverio/src/types.ts (hunk 1 above)
Comment (christian-bromann) [reply packages/webdriverio/src/types.ts:63]: 👍  let me take a look

**Comment by christian-bromann** [comment]
I finished the typing definition and refactored the typing tests to actually check for proper types. This is ready for review now.

**Pull request merged**